src/main/java/org/embulk/input/marketo/MarketoServiceImpl.java

```diff
@@ -0,0 +1,183 @@
+        marketoRestClient.startLeadBulkExtract(exportID);
+        try {
+            marketoRestClient.waitLeadExportJobComplete(exportID, pollingTimeIntervalSecond, bulkJobTimeoutSecond);
+        }
+        catch (InterruptedException e) {
+            LOGGER.error("Exception when waiting for export job id: {}", exportID, e);
+            throw new DataException("Error when wait for bulk extract");
+        }
+        InputStream leadBulkExtractResult = marketoRestClient.getLeadBulkExtractResult(exportID);
+        final File tempDir = Files.createTempDir();
+        TempFileSpace tempFiles = new TempFileSpace(tempDir);
+        File tempFile = tempFiles.createTempFile(DEFAULT_FILE_FORMAT);
+        try {
+            Files.write(ByteStreams.toByteArray(leadBulkExtractResult), tempFile);
+        }
+        catch (IOException e) {
+            LOGGER.error("Encounter exception when download bulk extract file, job id [{}]", exportID, e);
+            throw new DataException("Can't download bulk extract file");
+        }
+        return tempFile;
+    }
+
+    @Override
+    public File extractAllActivity(Date startTime, Date endTime, int pollingTimeIntervalSecond, int bulkJobTimeoutSecond)
+    {
+        String exportID = marketoRestClient.createActitvityExtract(startTime, endTime, null);
+        marketoRestClient.startActitvityBulkExtract(exportID);
+        try {
+            marketoRestClient.waitActitvityExportJobComplete(exportID, pollingTimeIntervalSecond, bulkJobTimeoutSecond);
+        }
+        catch (InterruptedException e) {
+            LOGGER.error("Exception when waiting for export job id: {}", exportID, e);
+            throw new DataException("Error when wait for bulk extract");
+        }
+        InputStream activitiesBulkExtractResult = marketoRestClient.getActivitiesBulkExtractResult(exportID);
+        final File tempDir = Files.createTempDir();
+        TempFileSpace tempFiles = new TempFileSpace(tempDir);
+        File tempFile = tempFiles.createTempFile(DEFAULT_FILE_FORMAT);
+        try {
+            Files.write(ByteStreams.toByteArray(activitiesBulkExtractResult), tempFile);
+        }
+        catch (IOException e) {
+            LOGGER.error("Encounter exception when download bulk extract file, job id [{}]", exportID, e);
+            throw new DataException("Can't download bulk extract file");
+        }
+        return tempFile;
+    }
+
+    @Override
+    public Iterable<ObjectNode> getAllListLead(List<String> fieldNames)
+    {
+        RecordPagingIterable<ObjectNode> lists = marketoRestClient.getLists();
+        List<Iterable<ObjectNode>> iterables = new ArrayList<>();
+        for (ObjectNode node : lists) {
+            final String id = node.get("id").asText();
+            iterables.add(Iterables.transform(marketoRestClient.getLeadsByList(id, FluentIterable.from(fieldNames).filter(new Predicate<String>()
```

Comment:
There are 2 issues with this line:
- It's long & complicated due to chaining several method invocations and cause reader hard to realize the purpose. If we can break it down to multiple lines, it will be great.
- The Function object can be refactored into a method to be reused in the next `getAllProgramLead` function.

Reply:
Fixed by move the the logic out of this class